### PR TITLE
Fixed windows bugs

### PR DESCRIPTION
**Pull request by JTvD**
Some of the features Christine implemented previously broke in the merge, like users which don't have acces to /home automatically ended up in /home/.
And ubuntu commands like 'which' not being available on a windows machine.

We also went through the pyinstaller script, the '.py' are only loaded if the application is frozen (build) so we don't need that if statement in the ui files anymore.

### CODE REVIEW

**Review by chStaiger — COMMENTED** [review]
not all users can access <zone>/home

How do users know which is their home? Do you then set the `irods_home` key in the irods_environment.json?
If so, this can be read in and used to get the collections a user has access to.

**Comment by jpmcfarland** [comment]
> not all users can access /home
> 
> How do users know which is their home? Do you then set the `irods_home` key in the irods_environment.json? If so, this can be read in and used to get the collections a user has access to.

The collection `/<zone name>/home` is not always configured readable for all users.  Quite often all users are part of a `public` group which has permission to see the contents of `/<zone name>/home`, but not always:

```
[irods@rdms ~]$ ils -A /rug/home  
/rug/home:
        ACL - g:public#rug:read object
```

We instruct users to add the `irods_home` variable to their `~/.irods/irods_environment.json` using their username as their home collection name:

```
"irods_home": "/rug/home/user.name@rug.nl",
```

I don't know if it can be depended on universally, but it is a place to start.  Perhaps look there first and fall back to constructed `/<zone name>/home/<user name>`, reminding the user to add this to their environment if neither is found.

**Comment by chStaiger** [comment]
> > not all users can access /home
> > How do users know which is their home? Do you then set the `irods_home` key in the irods_environment.json? If so, this can be read in and used to get the collections a user has access to.
> 
> The collection `/<zone name>/home` is not always configured readable for all users. Quite often all users are part of a `public` group which has permission to see the contents of `/<zone name>/home`, but not always:
> 
> ```
> [irods@rdms ~]$ ils -A /rug/home  
> /rug/home:
>         ACL - g:public#rug:read object
> ```
> 
> We instruct users to add the `irods_home` variable to their `~/.irods/irods_environment.json` using their username as their home collection name:
> 
> ```
> "irods_home": "/rug/home/user.name@rug.nl",
> ```
> 
> I don't know if it can be depended on universally, but it is a place to start. Perhaps look there first and fall back to constructed `/<zone name>/home/<user name>`, reminding the user to add this to their environment if neither is found.

I agree, check if `irods_home` is set and if not try `/zone/home/user`
I think this fix is not important for the merge. I will add this as a new issue to my git repo and label it with beta-release.

**Comment by chStaiger** [comment]
I think in this branch we do not have the extended info on modifying data any longer, the browser table changed back to the original state. I do not mind it, but are you ok with that @jpmcfarland ?

**Comment by jpmcfarland** [comment]
> I think in this branch we do not have the extended info on modifying data any longer, the browser table changed back to the original state. I do not mind it, but are you ok with that @jpmcfarland ?

If you are referring to the changes from PR #54, those are unaffected and will be integrated when that PR is merged.

@wurDevTim Please let me know when you are done with your changes and everything is ready for my review.

**Comment by JTvD** [comment]
Ehm I had hoped it was clear from the code but I see some explanation is needed. 

I dove into the <zone>/home  because the the servers at WUR are setup so that we cannot access this.  
During initialization the updownloadstab populates the irodsTreeView which crashed... 

The interesting this was that the old version, on Christine's git did work, so we looked at the differences and saw line 31-37 in https://github.com/chStaiger/iBridges-Gui/blob/master/gui/irodsTreeView.py
And as you also mention in the comments this fallback makes sense to have anyway, so we added it. 
But left the todo. 

When the documentation/tutorials are written I agree that it makes sense to use the `irods_home` in the `~/.irods/irods_environment.json` and implement this. But right now its not used, see line 57 at https://github.com/rijksuniversiteit-groningen/iBridges/blob/develop/gui/irodsTreeView.py

File: utils/IrodsConnector.py
Comment (jpmcfarland) [inline utils/IrodsConnector.py]: Use of the `uid` is no longer necessary when `python_irodsclient==1.1.6` is used.  The PR (#67) for that is pending review.  This dependency can even be updated here for consistency if desired for this change.

Comment (jpmcfarland) [reply utils/IrodsConnector.py]: Indeed, all UID-related code can be removed.  Those changes in the PR above tested correctly on Windows.

File: pyinstaller_script.py
Comment (jpmcfarland) [inline pyinstaller_script.py]: Why no more `sys.frozen` check?  And is `split("..")[1]` index always giving consistent results?

File: utils/IrodsConnector.py
Comment (JTvD) [reply utils/IrodsConnector.py]: When [67](https://github.com/rijksuniversiteit-groningen/iBridges/pull/67) is approved and merge, we can merge develop back into this branch and fix it, do you agree?

File: pyinstaller_script.py
Comment (JTvD) [reply pyinstaller_script.py]: No more `sys.frozen` becausse 'the '.py' are only loaded if the application is frozen (build) so we don't need that if statement in the ui files anymore.' As example: 
```
if getattr(sys, 'frozen', False):
    super().setupUi(self)
else:
    PyQt6.uic.loadUi("gui/ui_files/tabUpDownload.ui", self)
```

And yess, for as far as I know the  '..' operator specifies the parent folder on all operating systems, but I can currently only test on windows and ubuntu.

File: gui/IrodsUpDownload.py
Comment (jpmcfarland) [inline gui/IrodsUpDownload.py]: I think removing this causes the `LOCAL` FS tree view in this tab to default to the root entry, effectively `index = 0`.  The view in the ELabJournal tab is unaffected.

**Comment by jpmcfarland** [comment]
The view of the comments above are valid for both macOS and Windows.

**Review by chStaiger — APPROVED** [review]
I tested the main functionality. All fine for me.

**Comment by jpmcfarland** [comment]
PR #67 merged.   Only a single line conflict to be resolved.

File: utils/IrodsConnector.py
Comment (jpmcfarland) [reply utils/IrodsConnector.py]: Yes.  Only one line remains in conflict.

File: pyinstaller_script.py
Comment (jpmcfarland) [reply pyinstaller_script.py]: > And is split("..")[1] index always giving consistent results?

What I meant about consistent was, is there only ever 1 `..` in `icon_path`?  I remember previous examples where there are multiple such strings.  If the text following the last `..` is always the one you want, this will work best:

```python
                    icon_path = icon_path.split("..")[-1]
```

That will always give the _last_ element in the split list, **and** gives the correct result even if `..` is not in `icon_path`.

**Comment by jpmcfarland** [comment]
> Ehm I had hoped it was clear from the code but I see some explanation is needed.
> 
> I dove into the /home because the the servers at WUR are setup so that we cannot access this. During initialization the updownloadstab populates the irodsTreeView which crashed...
> 
> The interesting this was that the old version, on Christine's git did work, so we looked at the differences and saw line 31-37 in https://github.com/chStaiger/iBridges-Gui/blob/master/gui/irodsTreeView.py And as you also mention in the comments this fallback makes sense to have anyway, so we added it. But left the todo.
> 
> When the documentation/tutorials are written I agree that it makes sense to use the `irods_home` in the `~/.irods/irods_environment.json` and implement this. But right now its not used, see line 57 at https://github.com/rijksuniversiteit-groningen/iBridges/blob/develop/gui/irodsTreeView.py

I have been wanting for a while now to have the iRODS tree view open up by default to the `irods_cwd` environmental variable or to the home collection`/<zone_name>/home/<user_name>` if not defined (see Issue #4).  I think that is a far more ideal arrangement and far more iRODS-like, and solves this problem because a user now has better control over where they start in the tree views and browser tab.

File: pyinstaller_script.py
Comment (JTvD) [reply pyinstaller_script.py]: Yes, I understand. Unless we move the icons folder there will always be one and only one `..` in the path. But I agree that always getting the last element is even nicer. Updated this line.

File: utils/IrodsConnector.py
Comment (JTvD) [reply utils/IrodsConnector.py]: I merged the changes into this branch, there where a few small things but it works now.  And its really nice to login without having to retype the password every time.

**Comment by JTvD** [comment]
While I agree that it would be nice to have the tree's open on startup, that's a new feature and relatively small change. Not something this fix now during the merge. 

Reran all my tests successfully and checkout the fancy new browser. I think we got all the comments and changes now, let me know if I missed something?

File: utils/IrodsConnector.py
Comment (jpmcfarland) [reply utils/IrodsConnector.py]: Indeed!

File: pyinstaller_script.py
Comment (jpmcfarland) [reply pyinstaller_script.py]: Thanks!

**Comment by JTvD** [comment]
@jpmcfarland I saw you updated the comments, I don't see any more change requests. On this git I don't have clearance to merge branches, can you give it one final look and if you approve merge this pull request?

**Comment by jpmcfarland** [comment]
> @jpmcfarland I saw you updated the comments, I don't see any more change requests. On this git I don't have clearance to merge branches, can you give it one final look and if you approve merge this pull request?

@wurDevTim You missed the comment about the unset index for the local FS tree view in `IrodsUpDownload.py`.

File: gui/IrodsUpDownload.py
Comment (JTvD) [reply gui/IrodsUpDownload.py]: That is correct. When we are done with the merging we look into how to open both the local and irods file tree correctly. this is just a quick patch.

Comment (jpmcfarland) [reply gui/IrodsUpDownload.py]: Removing `self.localFsTreeView.setCurrentIndex(index)` causes the root of the tree in the data transfer tab to be different from the root in the ELN tab.  Shouldn't they be consistent?

Comment (JTvD) [reply gui/IrodsUpDownload.py]: Made the same changes in the elab tab and added todo's to discuss this in the future. If we indeed decide to make them similar we have some duplicate code to cleanup.

Comment (jpmcfarland) [reply gui/IrodsUpDownload.py]: I really thought the original view starting in the user's home directory was ideal and would like to extend that to the iRODS view.  The consistency I mentioned was between the local FS tree views in the different tabs, not the local FS and iRODS tree views.  Indeed, we can tackle that with Issue #4 or create a whole new issue for it.

Comment (JTvD) [reply gui/IrodsUpDownload.py]: While I agree that having the tree open would be nice, we have to modify the way it does this to make it robust. Same problem as I encountered for the irods path where '<Zone>/home' is not accessible for all users. 
Unfortunately I could not quickly find the right commands for it.
At WUR we currently don't use the ELN tab, therefore I had not noticed that it has the same issue. 

So my proposition would indeed be to define what to do in this case when we pickup [#4](https://github.com/rijksuniversiteit-groningen/iBridges/issues/4)

Comment (jpmcfarland) [reply gui/IrodsUpDownload.py]: I agree.  We will certainly tackle the iRODS tree view starting point at another moment.  We can even consider adding a configurable local FS tree view starting point as well.

But for now, is it an idea to return the local FS tree view starting points back to where they originally were?  Then at least current users will have to navigate only down the iRODS tree and not the local tree as well to find their respective homes when starting a new session.

Comment (JTvD) [reply gui/IrodsUpDownload.py]: If the original solution for the local FS tree view would work than I would not have removed it... 
If it's something you need quickly than it could be part of the next branch. 

But as this is an Alpha version of iBridges I don't see the problem with removing this feature now and making a fix in a new branch.

Comment (jpmcfarland) [reply gui/IrodsUpDownload.py]: I'm sorry, but it _does_ work.  What does removing this one line fix?  I really don't understand.  That's why I'm continuing this conversation.

This is my view now:
Screenshot 2023-02-22 at 10 24 42

This was my view before the change:
Screenshot 2023-02-22 at 10 29 14

I feel this original view far better until we can fix both views to our liking.

Comment (JTvD) [reply gui/IrodsUpDownload.py]: Your right, it works for you just like the 'which' command worked... but that does not mean that it works in general.
That's the frustrating part of multiplatform development. 
It has to work on a lot of devices and configurations and just like '<zone>/home' is not always accesable 'C:\Users\username' is also not always accessible from the GUI.  

But this is clearly a blocker for you, and were getting no where like this...
so I will put it back, we finish the merge and than I will fix it in a new branch when also looking at the other trees.

**Review by jpmcfarland — APPROVED** [review]
Looks great!  Thanks!

**Comment by jpmcfarland** [comment]
@wurDevTim Thanks!  Merge at your leisure.

**Comment by jpmcfarland** [comment]
May I merge this so I can get started with the upstream merge?

**Comment by JTvD** [comment]
yess please, if I could I would have pushed the button after you gave your approval.  
Lets finish the big merge so we can start the next development phase.

**Comment by jpmcfarland** [comment]
> yess please, if I could I would have pushed the button after you gave your approval.
> Lets finish the big merge so we can start the next development phase.

Hmm, I thought you should be able to.  I'll look into it.

**Pull request merged**
